OgreMain/src/OgreHlms.cpp

```diff
@@ -2141,51 +2145,206 @@ namespace Ogre
                        sizeof( "\n\tDONE DUMPING PIECES\n#endif\n" ) - 1u );
     }
     //-----------------------------------------------------------------------------------
-    void Hlms::applyStrongMacroblockRules( HlmsPso &pso )
+    void Hlms::applyStrongBlockRules( HlmsPso &pso )
     {
-        if( !pso.macroblock->mDepthCheck )
-        {
-            // Depth check is already off, we don't need to hold a strong reference.
-            pso.pass.strongMacroblockBits &= ~HlmsPassPso::NoDepthBuffer;
-        }
-        if( !pso.macroblock->mDepthWrite )
-        {
-            // Depth writes is already off, we don't need to hold a strong reference.
-            pso.pass.strongMacroblockBits &= ~HlmsPassPso::ForceDisableDepthWrites;
-        }
-        if( pso.macroblock->mCullMode == CULL_NONE )
-        {
-            // Without culling there's nothing to invert, we don't need to hold a strong reference.
-            pso.pass.strongMacroblockBits &= ~HlmsPassPso::InvertVertexWinding;
-        }
-        if( pso.macroblock->mDepthClamp )
+        assert( pso.strongBlocks = 0 );
+
+        // Modify a macroblock if needed.
+        HlmsMacroblock macroblock = *pso.macroblock;
+        macroblock.mRsData = nullptr;
+        mListener->applyStrongMacroblockRules( macroblock,
+                                               *this );  // Allows the listener to modify a macroblock
+        applyStrongMacroblockRules(
+            macroblock );  // Allows the implementation (inherited classes) to modify a macroblock
+        assert( macroblock.mRsData == nullptr );  // Check if implementation reassigned the macroblock
```

Comment:
Change:
`// Check if implementation reassigned the macroblock`
for:
`// Check implementation didn't reassign the macroblock.`